Speed up DOI extraction with new regexp

mudge opened the pull request:
Previously: running the DOI extraction test suite would take just over 2 seconds; with this new optimisation, it takes under 10 milliseconds.

The key is to replace the Ruby logic which would incrementally strip trailing punctuation off any matched DOI until it had a valid ending with a regular expression that has the same behaviour. This is possible if we're much more explicit about valid DOI suffixes.

mudge committed:
Speed up DOI extraction with new regexp

Previously: running the DOI extraction test suite would take just over 2
seconds; with this new optimisation, it takes under 10 milliseconds.

The key is to replace the Ruby logic which would incrementally strip
trailing punctuation off any matched DOI until it had a valid ending
with a regular expression that has the same behaviour. This is possible
if we're much more explicit about valid DOI suffixes.

ascastro commented on lib/identifiers/doi.rb:
```diff
@@ -1,43 +1,31 @@
     REGEXP = %r{
-      \b
-      10 # Directory indicator (always 10)
+      10                                      # Directory indicator (always 10)
```
By removing the `\b`, this will also match strings that don't start with 10. For example: `110.1097/01.ASW.0000443266.17665.19`.

mudge replied:
I'll add a test for it (this needs porting to the PHP version too as it was missing there).

ascastro commented on lib/identifiers/doi.rb:
```diff
@@ -2,42 +2,31 @@ module Identifiers
+          # DOI suffix
+          [^[:space:]]+2-\#                   # Early Wiley suffix
+          |
+          [^[:space:]]+\([^[:space:]]+\)      # Suffix ending in balanced parentheses
```
Actually - I found another problem. This will actually also match `10.1130/2013.2502(04))`. I think here, what we want is: `[^[:space:]]+\([^[:space:])]+\)`

mudge commented on lib/identifiers/doi.rb:
```diff
@@ -2,42 +2,31 @@ module Identifiers
+        /                                     # Prefix/suffix divider
+        (?:
+          # DOI suffix
+          [^[:space:]]+2-\#                   # Early Wiley suffix
```
Worth noting that I've gathered a sample of 600 DOIs from [Crossref's API](http://api.crossref.org/) and found more Wiley DOIs which end in a single number or letter so this probably needs expanding a little.

mudge committed:
Support wider variety of DOIs

By using Crossref's API, we collected a sample of nearly 1,000 DOIs and
tested our extraction against them. Aside from DOIs ending in full stops
(e.g. 10.17147/asui.), all tests now pass.

ascastro commented:
🎉
